src/EventStore.Core/TransactionLog/Chunks/TFChunk/TFChunk.cs

```diff
@@ -1066,8 +1105,36 @@ public enum ChunkVersions : byte {
 
 		private void ReturnReaderWorkItem(ReaderWorkItem item) {
 			if (item.IsMemory) {
+				// we avoid taking the _cachedDataLock here every time because we would be
+				// contending with other reader threads also returning readerworkitems.
+				//
+				// instead we check _cacheStatus to give us a hint about whether we need to lock.
+				// but this read of _cacheStatus is not inside the lock, so it can be wrong
+				// because we read a stale value, or it changed immediately after
+				// we read it.
+				//
+				// if the check is wrong it can result in one of two outcomes:
+				// 1. we call TryDestructMemStreams when we shouldn't.
+				//    we protect against this by checking the condition again inside the lock
+				// 2. we don't call TryDestructMemStreams when we should.
+				//    the memory barrier, corresponding to the barrier in UnCacheFromMemory, protects
+				//    against this by guaranteeing ordering as emphasied in _italics_ below.
+				//    the case protected against is that another thread might call UnCacheFromMemory just as
+				//    we are returning an item here and neither thread tidys up the item we are returning.
+				//    but consider: UnCacheFromMemory sets the state to Uncaching _and then_ tidys up the
+				//    items in the pool. if it does not tidy up our item, it is because our item wasn't
+				//    in the pool. which means we hadn't enqueued it yet, which means that we will
+				//    enqueue it and then _after that_ read _cacheStatus and find it is Uncaching,
```

Comment:
what guarantees do we have that the read of _cacheStatus will return `Uncaching` here? it appears that it may still be `Cached` but I may have gotten something wrong.

Reply:
after discussions with @timothycoleman and some research: a barrier should enforce cache coherence, so the above read should return the latest value given that we know it's already been written by another CPU core.